eventlet/dagpool.py

```diff
@@ -0,0 +1,484 @@
+
+        After a post(key, value1) followed by post(key, value2, replace=True),
+        it is unspecified which pending wait_each([key...]) calls (or
+        greenthreads iterating over 'results' involving that key) will observe
+        value1 versus value2. It is guaranteed that subsequent
+        wait_each([key...]) calls (or greenthreads spawned after that point)
+        will observe value2.
+        """
+        # First, check if we're trying to post() to a key with a running
+        # greenthread.
+        try:
+            coro = self.coros[key]
+        except KeyError:
+            # If there's no running greenthread with this key, carry on.
+            pass
+        else:
+            # A DAGPool greenthread is explicitly permitted to post() to its
+            # OWN key.
+            if coro.greenthread is not greenthread.getcurrent():
+                # oh oh, trying to post a value for running greenthread from
+                # some other greenthread
+                raise Collision(key)
+
+        # Here, either we're posting a value for a key with no greenthread or
+        # we're posting from that greenthread itself.
+
+        # Has somebody already post()ed a value for this key?
+        try:
+            self.values[key]
+        except KeyError:
+            # no, we're good
+            pass
+        else:
+            # unless replace == True, this is a problem
+            if not replace:
+                raise Collision(key)
+
+        # Either we've never before posted a value for this key, or we're
+        # posting with replace == True.
+
+        # update our database
+        self.values[key] = value
+        # and wake up pending waiters
+        self.event.send()
+        # The comment in Event.reset() says: "it's better to create a new
+        # event rather than reset an old one". Okay, fine. We do want to be
+        # able to support new waiters, so create a new Event.
+        self.event = Event()
+
+    def __getitem__(self, key):
+        """
+        __getitem__(key) (aka dagpool[key]) blocks until 'key' has a value,
+        then delivers that value.
+        """
+        # This is a degenerate case of wait_each(). Construct a list
+        # containing only this 'key'. list(wait_each()) will make a list
+        # containing exactly one (key, value) pair. Extract that pair, then
+        # extract just its value.
+        return list(self.wait_each([key]))[0][1]
+
+    def get(self, key, default=None):
+        """
+        get() returns the value for 'key'. If 'key' does not yet have a value,
+        get() returns 'default'.
+        """
+        return self.values.get(key, default)
+
+    def keys(self):
+        """
+        Return the list of keys for which we currently have values. Explicitly
+        return a copy rather than an iterator: don't assume our caller will
+        finish iterating before new values are posted.
+        """
+        return self.values.keys()
```

Comment:
Again, Python2-3 different semantics for `dict.keys`, `dict.items`, `dict.values`.